kasparovsyncd/sync.go

```diff
@@ -235,238 +255,594 @@ func doesBlockExist(dbTx *gorm.DB, blockHash string) (bool, error) {
-				}
-			}
-			missingParents = append(missingParents, parentHash)
-		}
-		return errors.Errorf("some parents are missing for block %s: %s", verboseBlock.Hash, strings.Join(missingParents, ", "))
+
+	addressToAddressID := make(map[string]uint64)
+	for _, dbAddress := range dbAddresses {
+		addressToAddressID[dbAddress.Address] = dbAddress.ID
 	}
 
-	for _, dbParent := range dbParents {
-		dbParentBlock := dbmodels.ParentBlock{
-			BlockID:       dbBlock.ID,
-			ParentBlockID: dbParent.ID,
+	newAddresses := make([]string, 0)
+	for address, id := range addressToAddressID {
+		if id != 0 {
+			continue
 		}
-		dbResult := dbTx.Create(&dbParentBlock)
-		dbErrors := dbResult.GetErrors()
-		if httpserverutils.HasDBError(dbErrors) {
-			return httpserverutils.NewErrorFromDBErrors("failed to insert parentBlock: ", dbErrors)
+		newAddresses = append(newAddresses, address)
+	}
+
+	addressesToAdd := make([]interface{}, len(newAddresses))
+	for i, address := range newAddresses {
+		addressesToAdd[i] = &dbmodels.Address{
+			Address: address,
 		}
 	}
-	return nil
-}
 
-func insertRawBlockData(dbTx *gorm.DB, rawBlock string, dbBlock *dbmodels.Block) error {
-	blockData, err := hex.DecodeString(rawBlock)
+	err := bulkInsert(dbTx, addressesToAdd, insertChunkSize)
 	if err != nil {
-		return err
-	}
-	dbRawBlock := dbmodels.RawBlock{
-		BlockID:   dbBlock.ID,
-		BlockData: blockData,
+		return nil, err
 	}
-	dbResult := dbTx.Create(&dbRawBlock)
-	dbErrors := dbResult.GetErrors()
+
+	var dbNewAddresses []*dbmodels.Address
+	dbResult = dbTx.
+		Where("address in (?)", newAddresses).
+		Find(&dbNewAddresses)
+	dbErrors = dbResult.GetErrors()
 	if httpserverutils.HasDBError(dbErrors) {
-		return httpserverutils.NewErrorFromDBErrors("failed to insert rawBlock: ", dbErrors)
+		return nil, httpserverutils.NewErrorFromDBErrors("failed to find blocks: ", dbErrors)
 	}
-	return nil
+
+	if len(dbNewAddresses) != len(newAddresses) {
+		return nil, errors.New("couldn't add all new addresses")
+	}
+
+	for _, dbNewAddress := range dbNewAddresses {
+		addressToAddressID[dbNewAddress.Address] = dbNewAddress.ID
+	}
+	return addressToAddressID, nil
 }
 
-func insertSubnetwork(dbTx *gorm.DB, transaction *rpcmodel.TxRawResult, client *jsonrpc.Client) (*dbmodels.Subnetwork, error) {
-	var dbSubnetwork dbmodels.Subnetwork
+type txWithMetaData struct {
+	verboseTx *rpcmodel.TxRawResult
+	id        uint64
+	isNew     bool
+	mass      uint64
+}
+
+func transactionIDtoTxWithMetaDataToTransactionIDs(transactionIDtoTxWithMetaData map[string]*txWithMetaData) []string {
+	transactionIDs := make([]string, len(transactionIDtoTxWithMetaData))
+	i := 0
+	for txID := range transactionIDtoTxWithMetaData {
+		transactionIDs[i] = txID
+		i++
+	}
+	return transactionIDs
+}
+
+func insertBlocksTransactions(dbTx *gorm.DB, client *jsonrpc.Client, blocks []*rawAndVerboseBlock) (map[string]*txWithMetaData, error) {
+	subnetworkIDToID, err := insertBlocksSubnetworks(dbTx, client, blocks)
```

Comment:
I think maybe insertBlocksSubnetworks should be outside insertBlocksTransactions.
They are releated, but not enought to be one-part-of-the-other IMHO.